File: 297-SerializeAndDeserializeBinaryTree/297-SerializeAndDeserializeBinaryTree.cpp
// Last updated: 2/19/2026, 1:35:07 PM
/**
* Definition for a binary tree node.
* struct TreeNode {
*     int val;
*     TreeNode *left;
*     TreeNode *right;
*     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
* };
*/
class Codec {
public:
 
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
 
        string ans1;
        shelper(root, ans1);
    
        return ans1;
 
    }
 
    void shelper(TreeNode* tn, string& ans) {
        if(tn){
            ans += to_string(tn->val);
            ans += " ";
            
            shelper(tn->left, ans);
            shelper(tn->right, ans);
 
        
        }
        else{
            ans += "n ";
            
        }
   
       
    }
 
    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        
        int index = 0;
        string current = extractor(data, index);
   
        TreeNode* root= dhelper(data, index, current);
        
        return root;
 
    }
 
    TreeNode* dhelper(string& data, int& index, string current){
        if(current != "n"){
            TreeNode* root = new TreeNode;
            root->val = stoi(current);
            
            string leftchild = extractor(data, index);
            
 
            
 
            root->left=dhelper(data, index, leftchild);
            string rightchild = extractor(data, index);
            root->right=dhelper(data, index, rightchild);
            return root;
 
        }
       return nullptr;
 
        
    }
 
    string extractor(string& data, int& index){
        string answer;
        while (data[index] != ' '){
            answer += (data[index]);

            index++;
        }
        index++;
        return answer;
    }
 
 
    
};
 
// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));

File: 297-SerializeAndDeserializeBinaryTree/297-SerializeAndDeserializeBinaryTree.py
# Last updated: 2/19/2026, 1:35:09 PM
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Codec:

    def serialize(self, root: TreeNode) -> str:
        result = []
        self._dfs_serialize(root, result)
        return "".join(result)

    def _dfs_serialize(self, node: TreeNode, result: list) -> None:
        if not node:
            result.append("n,")
            return
        
        result.append(str(node.val) + ",")
        self._dfs_serialize(node.left, result)
        self._dfs_serialize(node.right, result)

    def deserialize(self, data: str) -> TreeNode:
        self.index = 0
        return self._dfs_deserialize(data)

    def _dfs_deserialize(self, data: str) -> TreeNode:
        # Null marker
        if data[self.index] == 'n':
            self.index += 2  # skip "n,"
            return None

        # Parse integer value
        sign = 1
        if data[self.index] == '-':
            sign = -1
            self.index += 1

        num = 0
        while data[self.index] != ',':
            num = num * 10 + (ord(data[self.index]) - ord('0'))
            self.index += 1

        self.index += 1  # skip comma

        node = TreeNode(sign * num)
        node.left = self._dfs_deserialize(data)
        node.right = self._dfs_deserialize(data)
        return node

        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))